### DISCUSSION TITLE
Strategies for reducing data transfer from remote parquet?

### DISCUSSION THREAD

**Discussion by bmschmidt** [duckdb/duckdb-wasm#407] [General]
Thanks again for your work on this.

I'm curious if there are strategies for reducing the browser payload on remote parquet files. Putting this in 'discussions' because it's partly incumbent on the person uploading the data, although I wonder if duckdb-wasm can do more.

To test what's possible with partial queries on remote files, I've set up a static site (ngrams dot benschmidt dot org) that queries a 1GB parquet file with two columns: the words and counts for unigrams in one of the google ngrams databases. https://books.google.com/ngrams. 

Thanks to the recent fixes it runs pretty fast and caches parts of the database locally.

But the payload being sent to the browser is significantly larger than I'd expect. I'm searching by `word0`, which is sorted alphabetically, so that parquet metadata can indicate what chunks need to be read, and I've written the data in parquet with 130KB chunks to try to minimize transfer needs. But almost every transfer ends with a 16MB chunk; running three or four queries takes 60 MB or so.

I don't fully understand parquet metadata, but I thought that it should be possible to check whether chunks are relevant without reading so many of them in. Back of the envelope, all that really *needs* to happen for this to be read is probably 1MB or so in range indexes and at most 4 parquet chunks of actual data. Any thoughts on how to get down towards that floor?

**Comment by ankoh** [comment]
Interesting, I think we might have overshot the goal with the changes in https://github.com/duckdb/duckdb-wasm/issues/381.
We changed the readahead acceleration to a factor of 8 so reading 2-3 of your 130KB chunks sequentially will already max out at a readahead of 16MB.
In your case, a smaller factor might have worked much better.
So I think we're currently on a bumpy road towards more robust defaults but we might benefit a lot from making the readahead acceleration configurable per database and per file.
That way, we can iterate on that magic number a bit faster.

Readaheads are unfortunately our best bet right now since we're still tied to blocking query execution within the web worker.
Eventually, we want to reduce the readahead acceleration in favor of multiple concurrent fetches but this requires more ground-work towards asynchronous I/O in DuckDB.

**Reply by ankoh** [reply]
Picking up the second half of what you said, I think a very interesting opportunity for the future are remote reads of self-contained database files (e.g. duckdb and sqlite). If they have persisted indexes, we can just jump around e.g. within the btree using range requests, which would certainly speed up your scenario even further.

**Reply by bmschmidt** [reply]
Thanks for your response, makes sense, and I certainly see why close fixation on this should wait until the system can shift more to promises. But being able to fiddle based on knowledge of the format--in this case, probably capping the max transfer size somewhere around the block size--would be fun.

For what it's worth I think I was seeing fairly large transfers under the old system, too. I may experiment with significantly smaller chunk sizes in the parquet--it would reduce the compression ratio, but maybe not too dramatically.

If you'd indulge my curiosity, what are the difference would be between remote duckdb and remote parquet files? I had the impression that the general file layout differed mostly in the details (e.g. parquet having more elaborate compression, which is bad for disk reads but might be better over the web?)

**Reply by bmschmidt** [reply]
(In case it's useful I'm considering trying to talk about possibilities for this formats w/ people at NYU's data repository services for--say--deposits of data publications associated with a research paper to allow some on-the-fly access and column summaries. While they might be persuaded on parquet as an acceptable long-term storage/preservation format instead of zipped JSON/csv, specific database formats would never fly in that use case.)

**Reply by domoritz** [reply]
That makes sense. The DuckDB file format has changed a few times already and as far as I know there is no promise that it will be stable at any point.

**Reply by ankoh** [reply]
That's exciting! I'd also lean towards parquet in your case.

Regarding database files, Parquet does not contain persistent indexes and will therefore struggle with very selective queries.
DuckDB will happily skip entire row groups based on their min-max statistics but will usually still do a table scan.
Just open our web shell and check the query plan with `EXPLAIN`.

A persistent index similar to SQLites Btrees would allow us to perform point/range lookups using 3-4 (cached) tree pages which would significantly speed up selective queries over the network.
In the end, this depends on the workload.
Analytical queries usually favour table scans over indexes but we might sometimes lean more towards (remote) indexes in the browser to minimise the amount of fetched data.
This is still more on the vision side since DuckDB does not persist its ART indexes today but this will eventually change.
I think that remote reading auxiliary data structures might go a long way for DuckDB-Wasm.

**Reply by ankoh** [reply]
But the parquet file format is mature and will already be much faster to process than (compressed) json, so using it for the data repositories sounds fantastic!

**Reply by ankoh** [reply]
By the way, if you want to play around with an unpolished feature of the web shell, you can get information about the accessed file ranges using:

<img width="1472" alt="Screenshot 2021-11-20 at 01 38 19" src="https://user-images.githubusercontent.com/3986510/142707639-4ccc4ab0-02c6-4a08-b3b3-ed44ac7de279.png">

**Reply by ryan-williams** [reply]
At [shell.duckdb.org](https://shell.duckdb.org) today I see:
```
duckdb> .fstats
Unknown command: .fstats
```
`.help` shows:
```
.files track $FILE     Collect file statistics.
```
but I'm so far unable to get it to work as shown above.

Here's an example shell session attempt where I attempt to run `.files track`, a query, and `.files read` on a `.parquet` file:

<details><summary>Output</summary>

```
DuckDB Web Shell
Database: v0.9.2
Package:  @duckdb/duckdb-wasm@1.28.1-dev80.0

Connected to a local transient in-memory database.
Enter .help for usage hints.

duckdb> .files
No files registered

duckdb> .files track 's3://duckdb-wasm-test/1e5.parquet';
Tracking file statistics for: 's3://duckdb-wasm-test/1e5.parquet';

duckdb> .files
No files registered

duckdb> select year from 's3://duckdb-wasm-test/1e5.parquet' limit 10;
┌──────┐
│ year │
╞══════╡
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
│ 2001 │
└──────┘

duckdb> .files
┌───────────────────────────────────┬───────────┬──────────┬────────────┐
│ File Name                         ┆ File Size ┆ Protocol ┆ Statistics │
╞═══════════════════════════════════╪═══════════╪══════════╪════════════╡
│ s3://duckdb-wasm-test/1e5.parquet ┆ 4.43 MB   ┆ Buffer   ┆ false      │
└───────────────────────────────────┴───────────┴──────────┴────────────┘

duckdb> .files reads 's3://duckdb-wasm-test/1e5.parquet';
                                                                                                                                         █ 1 B   
┌───────────────────────────────────────────────┬───────────────────────────────────────────────┬───────────────────────────────────────────────┐
│                                               │                                               │                                               │
└───────────────────────────────────────────────┴───────────────────────────────────────────────┴───────────────────────────────────────────────┘
 Cold                                            Read-Ahead                                      Buffered                                       
 0 B                                             0 B                                             0 B                                            

duckdb> .files track 's3://duckdb-wasm-test/1e5.parquet';
Tracking file statistics for: 's3://duckdb-wasm-test/1e5.parquet';

duckdb> .files
┌───────────────────────────────────┬───────────┬──────────┬────────────┐
│ File Name                         ┆ File Size ┆ Protocol ┆ Statistics │
╞═══════════════════════════════════╪═══════════╪══════════╪════════════╡
│ s3://duckdb-wasm-test/1e5.parquet ┆ 4.43 MB   ┆ Buffer   ┆ false      │
└───────────────────────────────────┴───────────┴──────────┴────────────┘

duckdb>
```
</details> 

It shows 0 bytes of reads; suggestions welcome.

**Reply by chriszrc** [reply]
Any update here? Using `.fstats` at https://shell.duckdb.org/ still results in `Unknown command: .fstats`

**Comment by chriszrc** [comment]
@ankoh Now that duckdb can persist the art indexes:

https://duckdb.org/2022/07/27/art-storage.html

does this change the situation for querying remote parquet (say from wasm)?

**Reply by ryan-williams** [reply]
At a minimum, I hoped the new indexes would support more efficient queries against `.duckdb` files where `INDEX`es had been added. I tried to measure it and got scattered results: #1575. Data transferred still seems to be far higher than should be required.
